### DISCUSSION TITLE
How create custom activity with IActivityTypeProvider

### DISCUSSION THREAD

**Discussion by FarahmandMhmd** [elsa-workflows/elsa-core#1650] [General]
Hello , I want to create custom activity with IActivityTypeProvider but somethings are not clear to me.
consider the following two samples:

**Sample-1-**  add a type to custom activity (for example **SendHttpRequest** is an Elsa build-in activity)

![Screenshot (13)](https://user-images.githubusercontent.com/80684177/137621036-e4383dd6-bda7-4e35-b749-9380814b9e61.jpg)
![Screenshot (12)](https://user-images.githubusercontent.com/80684177/137621142-9d1a31f5-d476-4832-a8a8-679d4fde6fe7.jpg)

**MyActivity** gets  inputs , outputs and all properties of  SendHttpRequest!!!!! .
in other words , MyActivity will be a copy of SendHttpRequest that we can't extend it .

when i trigger workflow , it works fine same as SendHttpRequest.

_______________________________________________________________________________________________________________________

**Sample-2-**  No type for activity

![Screenshot (14)](https://user-images.githubusercontent.com/80684177/137621949-e9b7b298-9d90-41b5-bd10-21bbdeaaf711.jpg)

![Screenshot (15)](https://user-images.githubusercontent.com/80684177/137621954-3f6a276a-da2a-480b-acdf-d26e154ebb96.jpg)

**MyActivity** gets our inputs , outputs and all properties.

But the problem is when i trigger workflow ,  ExecuteAsync can't run. 

what is missing here?

**Questions:**
1- Is it possible to extend existing Elsa activity like SendHttpRequest in sample-1? if yes , How?
2- Is it necessary to set a type for MyActivity in sample-2?
3- How can i run ExecuteAsync of MyActivity in sample-2? 
4- Can anyone provide me with a sample that explains how to create custom activity with IActivityTypeProvider?

**Comment by sfmskywalker** [comment]
1. Should be possible. The "how" depends on what you want to do when extending this activity. For example, do you want to add more properties, rename properties, change its behavior entirely?
2. This is necessary. But this can be any type you want as long as it implements `IActivity`. This is necessary in order to be able to associate metadata. In a future version, we will try to drop this and allow you to use truly dynamic activity types.
3. You should not try to run `ExecuteAsync` yourself - this is done by the system.
4. There are currently 6 implementations of `IActivityTypeProvider` in the source code that I recommend you take a look at.

Let me know if you have any more questions about implementing your own activity type provider. It can be a bit tricky, but enables powerful scenarios.

**Comment by FarahmandMhmd** [comment]
Thanks for your reply @sfmskywalker
Let me explain more details about my scenario.

**Scenario :**  
**Create** - Create custom activity for each item in database. 
**Call** - They have to call an api by its own url and method (same as **SendHttpRequest**). 
**Hide**- They have to hide properties form user in designer like a black box!.
**Icon** - They should have a unique icon.

For example one of items is as below:
**Item_1** = {
Name = "Confirm Order" , // name of activity
Description= "call api and Confirm the order",
Category = Order, // category of activity in designer
Icon= "icon-Confirm", // set icon for activity in designer by custom plugin
Method = "Post" , // method for calling api
Url= "www.example.com/api/order/Confirm " // url for calling api
}

### Which of the following three implementations is correct for this scenario?


### **Implementation-1**

![A (2)](https://user-images.githubusercontent.com/80684177/139637176-e63243ac-17c2-4737-a14e-aad09e39464c.jpg)

I created new activity class called **ApiCall** based on  **SendHttpRequest** and removed its properties. 
set url and method in ApiCall activity **OnExecuteAsync**.

**Results**
- Create is ok
- call is ok
- hide is ok

**Problem**
Icons are not unique!
As I checked in elsa documentation , we can change the icon of activity with custom plugin. but problem is that activityIconProvider is set the icon for type.
elsaStudio.activityIconProvider.register(type , icon)
Therefore all custom activities will have same icon because they have same type as ApiCall !

**Questions:** 
1- Can we use activity name or other fields instead of Type to change icon?
2- When i set type in IActivityTypeProvider , for example in Implementation-1 ApiCall , custom activities will be exactly same as ApiCall with defferent name and description. I can't change it's properties , outputs , inputs via **DescribeAsync** in IActivityTypeProvider.
If i want to add another properties to them , i have to go to **ApiCall** class and add it .
For example if i add below code to this implementation , it will not added to activity properties.

 InputProperties = new ActivityInputDescriptor[]
                        {
                            new ActivityInputDescriptor
                            (
                                name:"ExtraProperty",
                                type: typeof(string),
                                uiHint: ActivityInputUIHints.SingleLine,
                                label: "ExtraProperty",
                                hint: "The URL to call api.",
                                defaultValue: "Extra",
                                isBrowsable: true
                            )
}



### **Implementation-2**
In this implementation i want to define everthing in IActivityTypeProvider.
I didn't define class that implements IActivity!!!! like ApiCall in previous sample.
for simplicity I removed foreach and used hard code to have only one custom activity.
also added two properties called URL and Method to custom activity with default values.

![139571266-44ae7794-8203-4dbc-aa7a-cb6a08a7f8f1](https://user-images.githubusercontent.com/80684177/139644702-59d38ea3-aa15-4ab2-90a6-4181977baa52.jpg)

**Results**
- Create is ok
- hide is ok

**Problem**
Icons are not unique!
failed to call api

**Questions:** 
1- what type should I choose in this scenario?
2- I set three break points that you can see them in code. How can hit them?
3- Is it possible to write some code in ExecuteAsync ? for example use HttpClient to send request and call api with url and method , something like what i did before in ApiCall OnExecuteAsync?
4- Does this scenario need to use another class like bookmarks or custom activity that implements IActivity?
5- As you replied me in discord , "ActivateAsync is a function that you need to provide in order to create an object of a class that implements IActivity" !
Do you mean that we can define custom activity in ActivateAsync with an object and no need to create external class that implements IActivity like ApiCall?



### **Implementation-3**

This Implementation is similar to Implementation-1 but only difference is using in-built Elsa activity **SendHttpRequest** as type.

![bbb](https://user-images.githubusercontent.com/80684177/139646762-259bed6c-dfc7-45f1-b1bc-e64be90ecfdf.jpg)

**Results**
- Create is ok

**Problem**
failed to call api.
failed to hide properties.
Icons are not unique!

**Questions:** 
1- Is it possible to hide properties of **SendHttpRequest** here?
As i explained in Implementation-1 , when i set type , i can't manipulate properties via **DescribeAsync**. it will not work.
2- How and where can i set url and method ? as i explained in Implementation-2 i can't hit the break point of ExecuteAsync.

**Comment by sfmskywalker** [comment]
All three implementations should work, but the best one to pursue in your case I think is **implementation 1**.
So let's dive into that.

I noticed one issue with your activity provider: when you create your `ActivityDescriptor`, you are using `"CallApi"` as the activity type name. This should be changed to `ca.TypeName`.

Next, let's go over each question:

1. Yes, you can determine the icon to display in whatever way you want. And given that your activity types are database-driven, you probably want to derive the icon to use from information stored in your `ActivityModel`.

To enable this, I just pushed an update that enables you to store this icon information together with the `ActivityDescriptor` you create by putting it in the new `CustomAttributes` dictionary. For example:

```csharp
var activityDescriptor = new ActivityDescriptor
{
   ...
   CustomAttributes = new { 
      ["Icon"] = ca.SomeSortOfIconInformation,
      ["Prototype"] = "CallApi" // Useful to use in client-side plugin to identify all variants of "CallApi" activities.
   }
};
```

Now you can write a custom client-side plugin for your custom activity that uses the "Icon" information to provide markup for the icon to display. For example, if your `SomeSortOfIconInformation` stored the full SVG markup, your plugin could use it like this:

```javascript
// A custom plugin for the CallApi activity.
function CallApiPlugin(elsaStudio) {
   const eventBus = elsaStudio.eventBus;

   // Handle certain events so we can customize the icon of the CallApi activity variants.
   eventBus.on(EventTypes.ActivityDesignDisplaying, context => {
      const activityDescriptor = context.activityDescriptor;

      // We're only interested in CallApi variants.
      if (activityDescriptor.customAttributes.Prototype !== 'CallApi')
        return;
  
      // Assign the Icon SVG markup to the activity icon field.
      context.activityIcon = activityDescriptor.customAttributes.Icon;
   });

   eventBus.on(EventTypes.ActivityDescriptorDisplayContext, context => {
      const activityDescriptor = context.activityDescriptor;

      // We're only interested in CallApi variants.
      if (activityDescriptor.customAttributes.Prototype !== 'CallApi')
        return;
  
      // Assign the Icon SVG markup to the activity icon field.
      context.activityIcon = activityDescriptor.customAttributes.Icon;
   });
}
``` 

You can of course DRY up the above code and reduce some of the duplicated code, but you get the idea: you get the custom Icon value stored in the custom attributes of the activity descriptor you provided via your custom activity type provider on the server, which is now available on the client side.

2. You can add, remove and modify activity properties as you want. In your example however, you are overwriting existing properties, which is why you only see the one property you created. If you want to add an extra property, you need to add it to the property list.

For example:

```csharp
DescribeAsync = () =>
{
   // Generate a descriptor using the `IDescribesActivityType` service so we get existing properties and their metadata:
   var activityDescriptor = await _describesActivityType.DescribeAsync(activityType, cancellationToken);

   // The activityDescriptor now holds all properties provided by `CalllApi`.
   // Add a property like this:
   var extraProperty = new ActivityInputDescriptor
   (
      name: "ExtraProperty",
      type: typeof(string),
      uiHint: ActivityInputUIHints.SingleLine,
      label: "ExtraProperty",
      hint: "The URL to call api.",
      defaultValue: "Extra",
      isBrowsable: true
    );

   activityDescriptor.InputProperties = activityDescriptor.InputProperties.Append(extraProperty).ToArray();
}
```

**Comment by FarahmandMhmd** [comment]
I don't know why it doesn't work when I change ActivityDescriptor **Type** to ca.Name.
I may have missed something or maybe there is a bug.

Items in database.
![Image 2](https://user-images.githubusercontent.com/80684177/139829596-3b6b4765-668c-40f1-9e12-4b3a9063b50a.jpg)

Implementation of IActivityTypeProvider
![Image 3](https://user-images.githubusercontent.com/80684177/139829935-2969abf8-4487-427b-8c91-9482fd3288c2.jpg)


In both cases (type = "CallApi" or type = ca.Name) , activities created as below :
![Image 8](https://user-images.githubusercontent.com/80684177/139835481-736d4e17-c808-47fc-9052-2c1ec2ff56c2.jpg)


If set ActivityDescriptor.Type to **"CallApi"** :
all properties of CallApi will be displayed in designer.
after trigger workflow , **OnExecutesAync** method of ApiCall will be executed and successfully call api.
![a](https://user-images.githubusercontent.com/80684177/139833592-6f4379a0-7f28-4e53-bd16-fa79913e6607.jpg)


If set ActivityDescriptor.Type to **ca.Name** , 
no properties and output will be displayed in designer . only default common and storage tab is there.
failed to execute  **OnExecuteAsync** method of ApiCall.

As you see in this image , there are no properties or output in this activity!.
![b](https://user-images.githubusercontent.com/80684177/139833623-792e1dd2-f86f-41d9-8cd7-0864e163ba15.jpg)
![Image 7](https://user-images.githubusercontent.com/80684177/139834185-5c2cb373-2ca0-4c32-ac42-0a0bb09bcaff.jpg)

**Reply by sfmskywalker** [reply]
Looking at your database records, don't use `Type` but `Name` - otherwise you will have duplicate activity type names (`CallApi`), while you want to have unique activity types (e.g. `PostActivity`).

**Reply by sfmskywalker** [reply]
Make sure to use the `Name` property for both the `TypeName` property of the `ActivityType` and the `Type` property of the `ActivityDescriptor`.

**Reply by FarahmandMhmd** [reply]
I didn't use Type that stored in database in the code!
please ignore it form database.
look at the code again.
I used **typeof(CallApi)** as type.
also **ca.Name** for value of **ActivityType.TypeName** and **ActivityDescriptor.Type**.

if you check your previous reply :
"I noticed one issue with your activity provider: when you create your ActivityDescriptor, you are using "CallApi" as the activity type name. This should be changed to ca.TypeName."

I fixed it , but it doesn't work!

But when I set the **ActivityDescriptor.Type** to **"CallApi"** , everything works fine except icons!

**Reply by sfmskywalker** [reply]
Sorry, I must have misread 😅 

It's hard to say why it doesn't work by looking at individual code snippets. All looks good at first glance, but clearly there's a detail missing somewhere or perhaps there's a little bug somewhere. I would need to see a full repro (sample project or your solution as-is that I can run on my machine) to understand the issue.

With regards to the icons not working, you haven't shown me any code that you tried based on my last response where I explained a possible implementation to getting custom icons displayed.

**Reply by FarahmandMhmd** [reply]
@sfmskywalker 
Did you check sample project in my github?

**Reply by sfmskywalker** [reply]
@FarahmandMhmd I just looked at your project, and now I understand why you weren't seeing any outcomes: when you create a custom activity descriptor, you also need to provide the available outcomes, like so:

```csharp
DescribeAsync = () =>
{
    var activityDescriptor = new ActivityDescriptor
    {
        Type = ca.Name,
        DisplayName = ca.Name,
        Category = ca.Category,
        Description = ca.Description,
        Outcomes = new[]{ OutcomeNames.Done } // Notice I added this line
    };
    return new ValueTask<ActivityDescriptor>(activityDescriptor);
}
```

With that, your activities will now display like this:

![image](https://user-images.githubusercontent.com/938393/142686479-6b8b5231-06b5-441e-a501-cf90d4574de2.png)

**Comment by FarahmandMhmd** [comment]
I still have not implement custom icons based on your suggestion. ( I will do that asap)

I used the power of **Razor** page to showing icons in scenario. 
I know my solution is not the best approach , but for now it is working. :-)

1- use **@inject** in  _Host.cshtm razor page to call services.
2- get items from database with service in _Host.cshtm
3- convert items (c# object) to javascript object
4-Import FontAwesome CDN.
5- set icons with custom plugin

My challenge was finding a way to convert items (c# object) to javascript object.
I reach below code after many searches and tests :-)
 **var activities = JSON.parse('@Html.Raw(Json.Serialize(activity))');**
I still can't understand it!!! (-:


![Image 9](https://user-images.githubusercontent.com/80684177/139862773-bbb1cd32-eeca-4a77-bc3b-fb5edb32f361.jpg)


This is the final result!

![Image 10](https://user-images.githubusercontent.com/80684177/139866751-1a8a1ccf-83d8-4efa-8e6c-d7fc200fabf9.jpg)

**Reply by sfmskywalker** [reply]
I like your solution! There's no reason you shouldn't do it this way: since you're using Razor, you can totally use its power to generate the appropriate JavaScript like you do 👍🏻

**Comment by FarahmandMhmd** [comment]
@sfmskywalker 
I pushed full project to CustomActivitySample repository in my github.
Please take a look at it whenever you have time!
Thank you very much for your help

**Reply by sfmskywalker** [reply]
@FarahmandMhmd I just looked at your project, and now I understand why you weren't seeing any outcomes: when you create a custom activity descriptor, you also need to provide the available outcomes, like so:

```csharp
DescribeAsync = () =>
{
    var activityDescriptor = new ActivityDescriptor
    {
        Type = ca.Name,
        DisplayName = ca.Name,
        Category = ca.Category,
        Description = ca.Description,
        Outcomes = new[]{ OutcomeNames.Done } // Notice I added this line
    };
    return new ValueTask<ActivityDescriptor>(activityDescriptor);
}
```

With that, your activities will now display like this:

![image](https://user-images.githubusercontent.com/938393/142686479-6b8b5231-06b5-441e-a501-cf90d4574de2.png)

**Reply by FarahmandMhmd** [reply]
@sfmskywalker Thanks for your reply.

I have tested this before. This only makes it look good in the designer , but after running workflow system.NullReferenceException will be thrown.

It only works properly when set the value of Type to "CallApi".

    var activityDescriptor = new ActivityDescriptor
    {
        Type = "CallApi",  
        DisplayName = ca.Name,
        Category = ca.Category,
        Description = ca.Description,
    };

**Reply by sfmskywalker** [reply]
Indeed, I see. To fix that, make sure to also add the following code to your activity type provider:

```csharp
ActivateAsync = context => ValueTask.FromResult<IActivity>(context.GetService<CallApi>()),
```

Full example for context:

```csharp
var activityType = new ActivityType
{
    TypeName = ca.Name,
    Type = typeof(CallApi),
    DisplayName = ca.Name,
    ActivateAsync = context => ValueTask.FromResult<IActivity>(context.GetService<CallApi>()), // Add this line.
    DescribeAsync = () =>
    {
        var activityDescriptor = new ActivityDescriptor
        {
            Type = ca.Name,
            DisplayName = ca.Name,
            Category = ca.Category,
            Description = ca.Description,
            Outcomes = new[] { OutcomeNames.Done }
        };
        return new ValueTask<ActivityDescriptor>(activityDescriptor);
    },
};
```

Admittedly, this is not very obvious based on the error message you would receive ("NullReferenceException"). I will see if I can improve this.

**Reply by sfmskywalker** [reply]
I just pushed a change that doesn't require you to provide a factory for `ActivityAsync` - instead, it will have a default lambda that does essentially the same thing as the line I added above. This offers users the option to provide their own factory if they want to, but in almost all cases you want to use the default factory anyway.
